### ISSUE TITLE
don't use rand(3) in C-core

### ISSUE THREAD

**Issue by markdroth** [grpc/grpc#17891]
`rand(3)` is not thread-safe, but we're using in several places in C-core.

It's not clear what the alternative is, since `rand_r(3)` may not be portable enough.

Note that the code in src/core/lib/backoff/backoff.cc does have a custom random number generator implementation, but I don't know whether it works for data types other than `double`.

**Comment by nicolasnoble** [comment]
`rand(3)` isn't threadsafe, in the sense it won't be reproducible with a given seen. But otherwise, it shouldn't corrupt things. If we don't care about the reproducibility of things, then it should be fine.

Worst-case scenario however, if this is because TSAN complains, I can re-license my [personal rand code](https://github.com/grumpycoders/uC-sdk/blob/master/libc/src/rand.c). It technically works on `uint32_t`.

The RNG function in backoff.cc looks a bit suspect to me, but it does actually work on `uint32_t` too. The magic number `1103515245` is not a prime; it's equal to 3^5*5*7*129749, so I am calling out the uniformity of the random distribution. The `double` conversion is done afterward, out of the generated uint32_t.


Long story short about RNG:

-) If you NEED reproducible RNG from a given seed, don't use `rand()`.
-) If you NEED secure, unpredictable RNG, don't use `rand()`.
-) If you NEED an RNG that has a proper, uniform distribution, don't use `rand()`.

Any other usage is fine.

**Comment by nicolasnoble** [comment]
Yeah, I just tried running the prng from backoff.cc, and it is deemed suspicious by [dieharder](https://webhome.phy.duke.edu/~rgb/General/dieharder.php) :-)

```
#=============================================================================#
#            dieharder version 3.31.1 Copyright 2003 Robert G. Brown          #
#=============================================================================#
   rng_name    |rands/second|   Seed   |
stdin_input_raw|  4.90e+07  |  63605812|
#=============================================================================#
        test_name   |ntup| tsamples |psamples|  p-value |Assessment
#=============================================================================#
   diehard_birthdays|   0|       100|     100|0.37640982|  PASSED
      diehard_operm5|   0|   1000000|     100|0.65765486|  PASSED
  diehard_rank_32x32|   0|     40000|     100|0.00000000|  FAILED
    diehard_rank_6x8|   0|    100000|     100|0.00000000|  FAILED
   diehard_bitstream|   0|   2097152|     100|0.00000000|  FAILED
        diehard_opso|   0|   2097152|     100|0.00000000|  FAILED
        diehard_oqso|   0|   2097152|     100|0.00000000|  FAILED
         diehard_dna|   0|   2097152|     100|0.00000000|  FAILED
diehard_count_1s_str|   0|    256000|     100|0.00000000|  FAILED
diehard_count_1s_byt|   0|    256000|     100|0.00000000|  FAILED
 diehard_parking_lot|   0|     12000|     100|0.00000000|  FAILED
    diehard_2dsphere|   2|      8000|     100|0.00000000|  FAILED
    diehard_3dsphere|   3|      4000|     100|0.00000000|  FAILED
     diehard_squeeze|   0|    100000|     100|0.00000000|  FAILED
        diehard_sums|   0|       100|     100|0.00000000|  FAILED
        diehard_runs|   0|    100000|     100|0.99107448|  PASSED
        diehard_runs|   0|    100000|     100|0.90072701|  PASSED
       diehard_craps|   0|    200000|     100|0.00000000|  FAILED
       diehard_craps|   0|    200000|     100|0.00000000|  FAILED
 marsaglia_tsang_gcd|   0|  10000000|     100|0.00000000|  FAILED
 marsaglia_tsang_gcd|   0|  10000000|     100|0.00000000|  FAILED
         sts_monobit|   1|    100000|     100|0.00000000|  FAILED
            sts_runs|   2|    100000|     100|0.00000000|  FAILED
          sts_serial|   1|    100000|     100|0.00000000|  FAILED
          sts_serial|   2|    100000|     100|0.00000000|  FAILED
          sts_serial|   3|    100000|     100|0.00000000|  FAILED
          sts_serial|   3|    100000|     100|0.00000026|  FAILED
          sts_serial|   4|    100000|     100|0.00000000|  FAILED
          sts_serial|   4|    100000|     100|0.00000000|  FAILED
          sts_serial|   5|    100000|     100|0.00000000|  FAILED
          sts_serial|   5|    100000|     100|0.00000000|  FAILED
          sts_serial|   6|    100000|     100|0.00000000|  FAILED
          sts_serial|   6|    100000|     100|0.00000000|  FAILED
          sts_serial|   7|    100000|     100|0.00000000|  FAILED
          sts_serial|   7|    100000|     100|0.00000000|  FAILED
          sts_serial|   8|    100000|     100|0.00000000|  FAILED
          sts_serial|   8|    100000|     100|0.00000000|  FAILED
          sts_serial|   9|    100000|     100|0.00000000|  FAILED
          sts_serial|   9|    100000|     100|0.00000000|  FAILED
          sts_serial|  10|    100000|     100|0.00000000|  FAILED
          sts_serial|  10|    100000|     100|0.00000000|  FAILED
          sts_serial|  11|    100000|     100|0.00000000|  FAILED
          sts_serial|  11|    100000|     100|0.00000000|  FAILED
          sts_serial|  12|    100000|     100|0.00000000|  FAILED
          sts_serial|  12|    100000|     100|0.00000000|  FAILED
          sts_serial|  13|    100000|     100|0.00000000|  FAILED
          sts_serial|  13|    100000|     100|0.00000000|  FAILED
          sts_serial|  14|    100000|     100|0.00000000|  FAILED
          sts_serial|  14|    100000|     100|0.00000000|  FAILED
          sts_serial|  15|    100000|     100|0.00000000|  FAILED
          sts_serial|  15|    100000|     100|0.00000000|  FAILED
          sts_serial|  16|    100000|     100|0.00000000|  FAILED
          sts_serial|  16|    100000|     100|0.00000000|  FAILED
         rgb_bitdist|   1|    100000|     100|0.00000000|  FAILED
         rgb_bitdist|   2|    100000|     100|0.00000000|  FAILED
         rgb_bitdist|   3|    100000|     100|0.00000000|  FAILED
         rgb_bitdist|   4|    100000|     100|0.00000000|  FAILED
         rgb_bitdist|   5|    100000|     100|0.00000000|  FAILED
         rgb_bitdist|   6|    100000|     100|0.00000000|  FAILED
         rgb_bitdist|   7|    100000|     100|0.00000000|  FAILED
         rgb_bitdist|   8|    100000|     100|0.00000000|  FAILED
         rgb_bitdist|   9|    100000|     100|0.00000000|  FAILED
         rgb_bitdist|  10|    100000|     100|0.00000000|  FAILED
         rgb_bitdist|  11|    100000|     100|0.00000000|  FAILED
         rgb_bitdist|  12|    100000|     100|0.00000000|  FAILED
rgb_minimum_distance|   2|     10000|    1000|0.00000000|  FAILED
rgb_minimum_distance|   3|     10000|    1000|0.00000000|  FAILED
rgb_minimum_distance|   4|     10000|    1000|0.00000000|  FAILED
rgb_minimum_distance|   5|     10000|    1000|0.00000000|  FAILED
    rgb_permutations|   2|    100000|     100|0.22610586|  PASSED
    rgb_permutations|   3|    100000|     100|0.90393543|  PASSED
    rgb_permutations|   4|    100000|     100|0.38282233|  PASSED
    rgb_permutations|   5|    100000|     100|0.09747166|  PASSED
      rgb_lagged_sum|   0|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|   1|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|   2|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|   3|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|   4|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|   5|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|   6|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|   7|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|   8|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|   9|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|  10|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|  11|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|  12|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|  13|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|  14|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|  15|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|  16|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|  17|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|  18|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|  19|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|  20|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|  21|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|  22|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|  23|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|  24|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|  25|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|  26|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|  27|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|  28|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|  29|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|  30|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|  31|   1000000|     100|0.00000000|  FAILED
      rgb_lagged_sum|  32|   1000000|     100|0.00000000|  FAILED
     rgb_kstest_test|   0|     10000|    1000|0.00000000|  FAILED
     dab_bytedistrib|   0|  51200000|       1|0.00000000|  FAILED
             dab_dct| 256|     50000|       1|0.00000000|  FAILED
Preparing to run test 207.  ntuple = 0
        dab_filltree|  32|  15000000|       1|0.59971282|  PASSED
        dab_filltree|  32|  15000000|       1|0.69214053|  PASSED
Preparing to run test 208.  ntuple = 0
       dab_filltree2|   0|   5000000|       1|0.00000000|  FAILED
       dab_filltree2|   1|   5000000|       1|0.00000000|  FAILED
Preparing to run test 209.  ntuple = 0
        dab_monobit2|  12|  65000000|       1|1.00000000|  FAILED
```

**Comment by ingako** [comment]
Thanks for working on this issue, we were stuck on the random behaviors of our program caused by this for almost a week :(

The use of thread-unsafe rand() is particularly problematic for people who rely on reproducible results for scientific research, and programs that rely heavily on the RNG being actually randomly distributed.

GRPC is a library that should sit inside other apps, so it could totally contaminate lots of other software and cause a lot of hard to diagnose downstream issues.

**Comment by peteroupc** [comment]
The problem with `rand` is not just that it's not thread-safe, but also that it uses an unspecified algorithm yet is seedable, as I [further explain](https://stackoverflow.com/a/52881465) in a Stack Overflow answer.

Is the backoff algorithm (or anywhere else where random numbers are used in GRPC) security sensitive?

If information security is not a concern in the backoff algorithm, an alternative is to have GRPC use a specific PRNG there.  And I list [several high-quality PRNG algorithms](https://github.com/peteroupc/peteroupc.github.io/blob/master/random.md#high-quality-prng-examples).

**Comment by markdroth** [comment]
The rand() behavior in the backoff code is just for the jitter part of the backoff calculation, so it shouldn't have any significant security implications.  I think our other usages are similar.

This is something that would be nice to fix, but I'm not aware of any practical problems that it's causing, so it hasn't been a high priority to work on.  @ingako, can you be more specific about what problem it caused you?

I'd be happy to accept a PR to fix this.

**Comment by ingako** [comment]
@markdroth our program calls rand(), but the results returned by rand() is not repeatable if the program calls grpc services. We had to replace all occurrences of rand() in our program with C++11's random library to bypass this issue.

I have made [a simple example](https://github.com/ingako/grpc-data-race) to demonstrate this issue.

**Comment by markdroth** [comment]
@ingako I'm sorry, but I still don't understand the practical problem here.  What behavior is gRPC providing that is not working for you?  Is this just "we have a policy against allowing rand(3) to be called in any of our code or any of our dependencies", or is there some behavior that is broken as a result of our use of rand(3)?

**Comment by ingako** [comment]
It is just the "policy against allowing rand(3) to be called in any of our code or any of our dependencies" is problematic for programs reply on repeatable rand() results.

**Comment by markdroth** [comment]
Okay, then I don't see a real functional problem here.  That means that we're not going to consider this high priority from our end.  But like I said before, we'd be happy to accept a PR to fix it if you'd like to contribute one.

**Comment by jtattermusch** [comment]
We acknowledge this to be a valid feature request but since no one has come forward to implement this, we are going to close it for now while maintaining the label/Help wanted to indicate that this can be picked up as future work.

**Comment by markdroth** [comment]
@veblush is looking into what's preventing us from using the absl random library.  See #23346.

**Comment by veblush** [comment]
Note that C++11 provides a better solution than c rand() so it's feasible to use it instead. ([ref](https://stackoverflow.com/questions/19665818/generate-random-numbers-using-c11-random-library)) Using abseil/random is easier to use, though.

**Comment by yashykt** [comment]
As noted on https://github.com/grpc/grpc/pull/39222, we want to build a SharedBitGen for this
